### PR TITLE
cryptsetup: create /run/generator

### PR DESCRIPTION
**Pull request by keszybz**
https://bugzilla.redhat.com/show_bug.cgi?id=1897802:
When booting from a luks-encrypted root fs systemd-cryptsetup issues the following warning:
systemd-cryptsetup[385]: WARNING: Locking directory /run/cryptsetup is missing!

Another option would be to create the directory from pid1, early on. Then
individual tools could assume it always exists. But that seems somewhat nicer
to only create from tools that will use it.

### CODE REVIEW

**Comment by yuwata** [comment]
nit: typo in commit message: `/run/generator` -> `/run/cryptsetup`

**Comment by keszybz** [comment]
Good point. I don't think it makes sense to push a change for this, since that'd trigger the CI again. We can do squash-and-merge.

**Comment by poettering** [comment]
Isnt this an implementation detail of libcryptsetup? I.e. shouldnt libcryptsetup create the dir itself when it sees it missing?

**Comment by bluca** [comment]
> Isnt this an implementation detail of libcryptsetup? I.e. shouldnt libcryptsetup create the dir itself when it sees it missing?

It does IIRC - but it complains loudly while doing so

**Comment by keszybz** [comment]
My thinking (which I should have put in the commit message…), was that it is reasonable for the library to not create the dir on its own: after all, it's just a library and it shouldn't manage "global" state on its own. Whether the presence of the directory is such global state or just an implementation detail is up for debate.

**Comment by poettering** [comment]
@mbroz what's your take on this?

> My thinking (which I should have put in the commit message…), was that it is reasonable for the library to not create the dir on its own: after all, it's just a library and it shouldn't manage "global" state on its own. Whether the presence of the directory is such global state or just an implementation detail is up for debate.

Well the library wants to put stuff below the dir, hence it better put the dir in place, too, no? i.e. if it is willing to own the stuff below the dir, it should also be willing to own the dir itself...

File: src/cryptsetup/cryptsetup.c [hunk 1]
```diff
@@ -1383,9 +1384,15 @@ static void remove_and_erasep(const char **p) {
                 log_warning_errno(r, "Unable to erase key file '%s', ignoring: %m", *p);
 }
 
+static void mkdir_run_cryptsetup(void) {
+        /* libcryptsetup expects the locking directory to exists. Let's create it if missing, but only if we
+         * are root. */
+        if (getuid() == 0)
```
Comment (poettering) [inline src/cryptsetup/cryptsetup.c:1390]: the tool only works as root...

Comment (mbroz) [reply src/cryptsetup/cryptsetup.c:1390]: For actiavation, yes, CAP_SYSADMIN is needed (strict requirement for device-mapper ioctl).
Anyway, if you cannot use temp files configuration here (too early?), I think this patch is ok.

Comment (poettering) [reply src/cryptsetup/cryptsetup.c:1390]: @mbroz the main question is though: why doesn't libcryptsetup create the dir on its own, when it notices it is missing and it needs to create a file in there? i.e. it could just do an mkdir() right before creating whatever it needs to create and ignore things on EEXIST

Comment (mbroz) [reply src/cryptsetup/cryptsetup.c:1390]: There is already a fallback that creates this dir in libcryptsetup (and prints the warning)
https://gitlab.com/cryptsetup/cryptsetup/-/blob/master/lib/utils_device_locking.c#L112
So the question is why this is not executed...

Comment (poettering) [reply src/cryptsetup/cryptsetup.c:1390]: Hmm, can't the warning be dropped? i.e. is there really reason to warn about this? maybe downgrade to debug level at least?

Comment (mbroz) [reply src/cryptsetup/cryptsetup.c:1390]: (If the question is just why it prints the warning, then we can add note it was created so it is more clear. But the warning is there to detect that it was created non-standard way, it is really just a fallback.)

Comment (poettering) [reply src/cryptsetup/cryptsetup.c:1390]: but why fallback, why not make it a fully supported way?

the thing is that our usual way to create runtime dirs like this is via tmpfiles, but that only works for later boot stuff really. livcryptsetup runs super-duper early though, in the initrd already, hence we'd have to manually work around for this. I'd much prefer if we'd just make it official that libcryptsetup creates the dir if it is missing.

tmpfiles would still be able to adjust access mode or so if necessary on the dir, once it runs, but i see no reason why libcryptsetup couldn't just prepare things automatically as necessary so that it is robust to just run any time, including in earliest boot.

Comment (mbroz) [reply src/cryptsetup/cryptsetup.c:1390]: If the later transititon to tmpfiles is safe, then why not (move the message to debug). @oniko ?

Comment (oniko) [reply src/cryptsetup/cryptsetup.c:1390]: @mbroz yes, we can make it debug message, it's just:

are we sure we don't hide some ugly issue by this? I don't understand systemd early boot internals good enough. If /run/cryptsetup directory is created exactly only once, it should be safe. But if there's possibility of having two or more locking directory instances it's roadway to metadata corruption report...

Comment (poettering) [reply src/cryptsetup/cryptsetup.c:1390]: > are we sure we don't hide some ugly issue by this? I don't understand systemd early boot internals good enough. If /run/cryptsetup directory is created exactly only once, it should be safe. But if there's possibility of having two or more locking directory instances it's roadway to metadata corruption report...

Should be safe. /run is mounted in the initrd first, without selinux. During the initrd→host transition the mount is moved to the new root, and thus survives in its entirety with all open files intact. During the transition after PID 1 loaded the selinux policy it will iterate through /run in its entirety and relabel things, so that they match the policy. Then, eventually `systemd-tmpfiles` will run, and create the dir if it is missing, and adjust perms/ownership if it already exists. And that's all.

So, the outcome whether tmpfiles creates it or libcryptsetup already created it first should be the same: the dir will exist, will be selinux handled, will have right perms access mode, and will per the very same dir since it was created at first regardless if that's in the initrd or the host, and all files and their fds will remain intact through the whole procedure.

TLDR: this should be safe. (and if there's a problem it would be up to systemd/dracut to fix it, not libcryptsetup)

Comment (mbroz) [reply src/cryptsetup/cryptsetup.c:1390]: ok, see https://gitlab.com/cryptsetup/cryptsetup/-/commit/1e7521c0564936fdbf098ce448e91e0ba25bffae
(I think this makes this PR obsolete :)

There will be soon stable cryptsetup minor release fixing problems in dm-integrity, so I will backport it there.

Comment (poettering) [reply src/cryptsetup/cryptsetup.c:1390]: excellent! thanks so much!
